### DISCUSSION TITLE
Add configuration for group representation in OIDC

### DISCUSSION THREAD

**Discussion by offby0point5** [kanidm/kanidm#1769] [Ideas]
I'm testing Kanidm for a smaller organization and it looks good so far.
Although I found the way of sharing group membership with resource providers to be inflexible, as just every groups UUID gets listed.

On Forgejo (fork of Gitea) OIDC is natively supported. One can map groups supplied by the identity provider to organization and team memberships. In this case one can just use the UUIDs that are needed and ignore all others.

On Nextcloud, I couldn't find a better way than using the Nextcloud app `user_oidc`, which supports adding users to the Nextcloud's groups based on the `groups` claim. But this app is missing the mapping feature. As the `groups` claim is just a list of UUIDs, those are used as group names. Additionally even `idm_all_accounts` and other groups that are not needed on the cloud instance are included and Nextcloud's OICD app adds them as user groups, so one ends up with 20 groups with nonsensical names (the UUIDs) for every user of which 5 are actually needed. This makes managing group permissions like group folder access unpleasant or even impossible for non-admins that cannot correlate UUIDs and group names using Kanidm's CLI.

I thought about options to switch an OAuth2 system to either share the groups names or their UUIDs, much like the `prefer-short-username` and `prefer-spn-username` options.
Maybe there could be a filter or mapping option to just expose certain groups to resource providers or even name them differently if needed.

As I don't have the deep understanding of Kanidm, I put this here to collect some ideas before filing an issue.

**Comment by yaleman** [comment]
The UUID-name part is up to the RP, looks like [at least one issue](https://github.com/nextcloud/user_oidc/issues/640) and a [PR](https://github.com/nextcloud/user_oidc/pull/626) has been logged in the `user_oidc` app repository about it.

**Comment by yaleman** [comment]
I'm not sure about limiting the groups, surely this'd be made neater once the group-name thing is cleaned up on the RP end?

**Reply by TheRealGramdalf** [reply]
**Edit:** Looks like this is on the roadmap: https://github.com/kanidm/kanidm/issues/2319

In my opinion it would be intuitive for groups/roles to function similar to scope mappings - an administrator must link a Kanidm group to a claim, which is in turn linked to a scope. This would (as far as I know) solve the UUID issue, since the group name is already scoped to a specific resource server. This would be more flexible, and (again, in my opinion) a sightly saner default - it adopts the idea of least privilege, requiring an admin to explicitly configure a group mapping, rather than listing every single group at once.

I could see this being a scaling issue as well - if Kanidm always gives every single UUID to the resource server, requests could get large the larger a deployment is. It isn't an issue in a self hosting environment, where you won't (or shouldn't) have a user in more than say 50 groups (to be generous), but could become an issue with very large deployments.

**Reply by Firstyear** [reply]
Yes, https://github.com/kanidm/kanidm/issues/2319 already covers this.

**Comment by Firstyear** [comment]
The reason for the uuids is to make sure there is a static never changing ID. This is more important for usernames IMO since they are more likely to need to change. 

I think for groups though, we could list both uuid and spn so that the consumer has both options.
